jcache/remote/src/main/java/org/infinispan/jcache/remote/RemoteCacheWithSyncListeners.java

```diff
@@ -0,0 +1,113 @@
+      super(delegate);
+      this.notifier = notifier;
+      this.cache = cache;
+      this.timeout = delegate.getRemoteCacheManager().getConfiguration().socketTimeout();
+   }
+
+   @Override
+   public V put(final K key, final V value) {
+      return withSyncListeners(notifier.hasSyncCreatedListener(), key, value, new Callable<V>() {
+         @Override
+         public V call() throws Exception {
+            return delegate.put(key, value);
+         }
+      });
+   }
+
+   @Override
+   public V putIfAbsent(final K key, final V value) {
+      return withSyncListeners(notifier.hasSyncCreatedListener(), key, value, new Callable<V>() {
+         @Override
+         public V call() throws Exception {
+            return delegate.putIfAbsent(key, value);
+         }
+      });
+   }
+
+   public V remove(final Object key) {
+      return withSyncListeners(notifier.hasSyncRemovedListener(), DONT_EXPECT_EVENT_ON_NULL_RESULT, (K) key, null, new Callable<V>() {
+         @Override
+         public V call() throws Exception {
+            return delegate.remove(key);
+         }
+      });
+   }
+
+   @Override
+   public V replace(final K key, final V value) {
+      return withSyncListeners(notifier.hasSyncUpdatedListener(), key, value, new Callable<V>() {
+         @Override
+         public V call() throws Exception {
+            return delegate.replace(key, value);
+         }
+      });
+   }
+
+   @Override
+   public boolean replaceWithVersion(final K key, final V newValue, final long version) {
+      return withSyncListeners(notifier.hasSyncUpdatedListener(), key, newValue, new Callable<Boolean>() {
+         @Override
+         public Boolean call() throws Exception {
+            return delegate.replaceWithVersion(key, newValue, version);
+         }
+      });
+   }
+
+   private <R> R withSyncListeners(boolean hasListeners, K key, V value, Callable<R> callable) {
+      return withSyncListeners(hasListeners, true, key, value, callable);
+   }
+
+   private <R> R withSyncListeners(boolean hasListeners, boolean expectEventOnNull, K key, V value, Callable<R> callable) {
+      try {
+         if (!hasListeners) {
+            return callable.call();
+         }
+         CountDownLatch latch = new CountDownLatch(1);
+         notifier.addSyncNotificationLatch(cache, key, value, latch);
+         try {
+            R ret = callable.call();
+            if ((ret == null) && !expectEventOnNull) {
+               return ret;
+            }
+            boolean wasClosed = latch.await(timeout, TimeUnit.MILLISECONDS);
+            if (!wasClosed) {
+               log.timeoutWaitingEvent();
+            }
+            return ret;
+         } finally {
+            notifier.removeSyncNotificationLatch(cache, key, value, latch);
+         }
+      } catch (Exception ex) {
+         if (!(ex instanceof CacheEntryListenerException)) {
```

Comment:
This is personal preference, but it seems if we just had 2 catch blocks where one is CacheEntryListenerException that is just rethrown and the other is Exception would be a little more clear.  Then no if statement is even required.